src/main/java/io/pravega/connectors/flink/source/enumerator/PravegaSplitEnumerator.java

```diff
@@ -0,0 +1,200 @@
+        LOG.info("Starting the PravegaSplitEnumerator for reader group: {}/{}.", this.scope, this.readerGroupName);
+
+        if (this.readerGroupManager == null) {
+            this.readerGroupManager = ReaderGroupManager.withScope(scope, clientConfig);
+        }
+        if (this.readerGroup == null) {
+            this.readerGroupManager.createReaderGroup(this.readerGroupName, readerGroupConfig);
+            this.readerGroup = this.readerGroupManager.getReaderGroup(this.readerGroupName);
+        }
+        if (this.checkpoint != null) {
+            LOG.info("Recover from checkpoint: {}", checkpoint.getName());
+            this.readerGroup.resetReaderGroup(ReaderGroupConfig
+                    .builder()
+                    .maxOutstandingCheckpointRequest(this.readerGroupConfig.getMaxOutstandingCheckpointRequest())
+                    .groupRefreshTimeMillis(this.readerGroupConfig.getGroupRefreshTimeMillis())
+                    .disableAutomaticCheckpoints()
+                    .startFromCheckpoint(this.checkpoint)
+                    .build());
+        }
+    }
+
+    @Override
+    public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+        // the Pravega source pushes splits eagerly, rather than act upon split requests
+    }
+
+    // check with the current parallelism and update the assignment (always one-to-one mapping)
+    @Override
+    public void addReader(int subtaskId) {
+        if (enumContext.registeredReaders().size() == enumContext.currentParallelism() && splits.size() > 0) {
+            LOG.debug("Adding reader {} to PravegaSplitEnumerator for reader group: {}/{}.",
+                    subtaskId,
+                    scope,
+                    readerGroupName);
+            int numReaders = enumContext.registeredReaders().size();
+            Map<Integer, List<PravegaSplit>> assignment = new HashMap<>();
+            for (int i = 0; i < numReaders; i++) {
+                assignment.put(i, Collections.singletonList(splits.get(i)));
+            }
+            enumContext.assignSplits(new SplitsAssignment<>(assignment));
+            splits.clear();
+        }
+    }
+
+    // deal with the data recovery, it will call readerGroup::initiateCheckpoint to get a checkpoint for recovery.
+    // this call will be handled in another thread pool instead of the split enumerator thread.
+    @Override
+    public Checkpoint snapshotState(long chkPtID) throws Exception {
+        final String checkpointName = createCheckpointName(chkPtID);
+
+        LOG.info("Initiate checkpoint {}", checkpointName);
+        final CompletableFuture<Checkpoint> checkpointResult =
+                this.readerGroup.initiateCheckpoint(checkpointName, scheduledExecutorService);
+        try {
+            this.checkpoint = checkpointResult.get(5, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            LOG.error("Pravega checkpoint met error.", e);
+            this.checkpoint = null;
```

Comment:
Is this the expected behavior in the case of failure, to set checkpoint to null rather than bubble the exception up?

Reply:
Yes we should bubble the exception up here rather than returning null and leave it for flink to handle it. 
Also `checkpoint` should not be set to null because it represents the 'latest' checkpoint of the enumerator, thus, should keep unchanged when initiating new checkpoint fails.